src/bt.rs

```diff
@@ -74,324 +73,217 @@ pub enum BinaryTreeError<V> {
-    /// a new node containing the value is inserted.
-    ///
-    /// # Returns
-    /// - `Ok(())` when the node is inserted at the end of the path.
-    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
-    /// value at the end of the path.
-    /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
-    ///
-    /// In the error cases, no insertion occurs and the value is returned to
-    /// the caller of this function.
-    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        if let Some(node) = &mut self.root {
-            node.insert(path, value)
-        } else if path.is_empty() {
-            self.root = Some(Box::new(Node::new(value)));
-            Ok(())
-        } else {
-            Err(BinaryTreeError::UnreachableNode(value))
+            let next = &self.nodes[node?];
+            node = if !bit { next.left } else { next.right };
         }
-    }
-
-    /// Gets a reference to the value located at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a reference to the
-    /// value stored in the node.
-    pub fn get(&self, path: &Path) -> Option<&V> {
-        self.root.as_ref().and_then(|node| node.get(path))
-    }
 
-    /// Gets a mutable reference to the node located at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a mutable reference
-    /// to the node.
-    pub fn get_node(&mut self, path: &Path) -> Option<&mut Node<V>> {
-        self.root.as_mut().and_then(|node| node.get_node(path))
+        Some(&mut self.nodes[node?])
     }
 }
 
 impl<V> Default for BinaryTree<V> {
     fn default() -> Self {
         Self {
             root: Option::default(),
+            nodes: Vec::with_capacity(Self::NODES_CAPACITY),
         }
     }
 }
 
-impl<V: Encode> Encode for Node<V> {
-    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
-        self.value.encode(bytes)?;
+// Indicates the datatype used to serialize usize values.
+type UsizeType = u16;
 
-        let mut stack = Vec::new();
-        stack.push(&self.right);
-        stack.push(&self.left);
-
-        // Nodes are stored following a pre-order traversal.
-        while let Some(elem) = stack.pop() {
-            match elem {
-                None => CodecMarker::Leaf.encode(bytes)?,
-                Some(node) => {
-                    CodecMarker::Inner.encode(bytes)?;
-                    node.value.encode(bytes)?;
-                    stack.push(&node.right);
-                    stack.push(&node.left);
-                }
-            }
+impl Encode for Option<usize> {
```

Comment:
Rather than defining `Encode` for `Option<usize>`, should we do this on an option of a newtype (perhaps named something like `Option<TreeIndex>`)? Otherwise, every user of the `Encode` trait will be locked into this implementation on `Option<usize>`, and IMO this is not a general-purpose implementation.

Reply:
This was removed as is now not needed with the new serialization method.